### PR TITLE
Store object states and diffs to audit event target 

### PR DESCRIPTION
**Pull request by nikomakela**
NOTE: Pipeline configs in https://dev.azure.com/City-of-Helsinki/kuva-notification-service-api/_git/kuva-notification-service-api-pipelines/pullrequest/10132

NS-18 NS-177.

Store obejct states to the target of an audit event.

Available object storing options are:
- Store none. Default
- Store old state only
- Store new state only
- Store both, the old and the new states
- Store only the diff of the object states

Add a new object state serializer that can be used as a helper tool to
create the object state objects that can then be used in the audit event
target.

Add the object states writing to
1. querysets in manages.py of audit log app
2. admin.py views of audit log app. (It is from there automatically
enabled to delivery log views)
3. generic view bases of the api app

### CODE REVIEW

**Comment by terovirtanen** [comment]
**Notification service branch is deployed to platta: https://kuva-notification-service-pr78.api.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**Notification service branch is deployed to platta: https://kuva-notification-service-pr78.api.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**Notification service branch is deployed to platta: https://kuva-notification-service-pr78.api.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**Notification service branch is deployed to platta: https://kuva-notification-service-pr78.api.dev.hel.ninja** :rocket::rocket::rocket:

File: api/views.py [hunk 1]
```diff
@@ -129,9 +131,10 @@ def delivery_log_webhook(request, id):
     # Write audit log of the action
     audit_log_service._commit_to_audit_log(
         message=create_api_commit_message_from_request(
-            request,
-            Operation.UPDATE.value,
-            [str(id)],
+            request=request,
+            operation=Operation.UPDATE.value,
+            object_ids=[str(id)],
+            old_objects=[log],
```
Comment (karisal-anders) [inline api/views.py:137]: Isn't the `log` here actually a new object? I mean `log.update_report(request.data)` above has already been run. Having the old object also would be nice as this is an update operation. I'm thinking how'd that be possible, a deep-copy perhaps?

Comment (nikomakela) [reply api/views.py]: changed to `new_objects`

**Comment by terovirtanen** [comment]
**Notification service branch is deployed to platta: https://kuva-notification-service-pr78.api.dev.hel.ninja** :rocket::rocket::rocket:

File: audit_log/tests/test_utils.py [hunk 2]
```diff
@@ -61,3 +72,148 @@ def test_get_response_status(status_code, audit_status):
...
     assert get_response_status(res_mock) == audit_status
+
+
+@pytest.mark.parametrize(
+    "old_dict, new_dict, expected_diff",
+    [
+        ({}, {}, {}),
+        ({"a": 1}, {"a": 1}, {}),
+        ({"a": 1}, {"a": 2}, {"a": 2}),
+        ({"a": 1}, {"b": 2}, {"b": 2}),
```
Comment (karisal-anders) [inline audit_log/tests/test_utils.py:83]: IMHO e.g. adding a note to the `diff_dicts` function's docstring that it won't return deleted keys in the output at all would make this behavior even more explicit beyond the current "Returns a new dictionary containing only the changed keys and values.".

**Review by karisal-anders — APPROVED** [review]
Looks like an improvement, although is the diff used anywhere currently? Old and new object states are used explicitly, but is the diff version used anywhere? Please take a look at the comments that I left, they're about adding explicitness to the code by adding some comments to it.

File: audit_log/admin.py [hunk 3]
```diff
@@ -31,7 +31,9 @@ def get_changelist_instance(self, request):
         page = paginator.get_page(page_number)
         current_page_queryset = page.object_list
         current_page_queryset.with_audit_log_and_request(
-            request=request, operation=Operation.READ.value
+            request=request,
+            operation=Operation.READ.value,
+            force_disable_object_states=True,
```
[the remark this answers is not in the document]
Comment (nikomakela) [reply audit_log/admin.py:37]: Added a comment

File: audit_log/tests/test_utils.py (hunk 2 above)
Comment (nikomakela) [reply audit_log/tests/test_utils.py:83]: A note added

**Comment by terovirtanen** [comment]
**Notification service branch is deployed to platta: https://kuva-notification-service-pr78.api.dev.hel.ninja** :rocket::rocket::rocket:

**Pull request merged**